File: src/lib/crm/projectStatistics.ts
import { MEASURE_DEFINITIONS, type MeasureDefinition } from "./measureDefinitions";

/**
 * The whole property in numbers — every figure a project reports, each one
 * carrying the definition it was computed under.
 *
 * Modelled on the reference's statistics sheet, which is the right idea: one
 * screen, every measurement, an ⓘ on each row. What differs is which figures
 * belong on it, because this is a restoration file rather than an appraisal.
 *
 * Two deliberate departures.
 *
 * **Volume is promoted, not buried.** The reference lists it last, as trivia.
 * Dehumidification is sized from cubic footage, so here it sits with the
 * measurements an estimate is actually built from.
 *
 * **The three ground-surface variants are NOT reproduced.** The reference
 * separates ground surface "with all walls", "with interior walls" and
 * "without walls" — figures that need the THICKNESS of every wall and which
 * side of it belongs to which room. A RoomPlan scan gives wall faces, not
 * wall assemblies; we do not know whether a partition is 2×4 or 2×6, and a
 * scan of one room cannot tell an exterior wall from a shared one. Computing
 * those three numbers would mean inventing a thickness and reporting the
 * result as measured. One honest floor area, defined, beats three fabricated
 * ones — so this reports the area to the scanned wall faces and says so.
 *
 * Platform-neutral on purpose: it takes plain numbers, not a database row or
 * a `SavedScan`, so the same computation serves the API, the report and the
 * phone without any of them reaching into the others.
 */

/** One room's contribution, in metres and square metres throughout. */
export type StatisticsRoom = {
  level: string;
  floorAreaSqm: number;
  perimeterM: number;
  ceilingHeightM: number;
  wallAreaGrossSqm: number;
  wallAreaNetSqm: number;
  doorCount: number;
  windowCount: number;
};

export type StatisticRow = {
  /** Stable key, for looking a row up rather than matching its label. */
  id: string;
  label: string;
  value: number;
  unit: "count" | "m" | "sqm" | "cbm";
  /** The definition behind the figure, where one exists. Counts have none —
      "how many doors" needs no defending. */
  meaning?: MeasureDefinition;
};

export type ProjectStatistics = {
  summary: StatisticRow[];
  measurements: StatisticRow[];
};

/**
 * Every figure for a set of rooms.
 *
 * Sums, never averages. A property's wall area is the sum of its rooms' wall
 * areas; there is no meaningful "average room". Ceiling height is the one
 * exception and it is deliberately absent from the totals — averaging the
 * heights of a basement and a stairwell produces a number that describes
 * neither, and volume is already the figure that height was needed for.
 */
export function projectStatistics(rooms: StatisticsRoom[]): ProjectStatistics {
  const sum = (pick: (room: StatisticsRoom) => number) =>
    rooms.reduce((total, room) => total + (Number.isFinite(pick(room)) ? pick(room) : 0), 0);

  // Distinct floors, not rooms-with-a-level: three rooms on the Ground floor
  // are one floor.
  const floors = new Set(rooms.map((room) => room.level).filter(Boolean)).size;

  return {
    summary: [
      { id: "floors", label: "Floors", value: floors, unit: "count" },
      { id: "rooms", label: "Rooms", value: rooms.length, unit: "count" },
      { id: "doors", label: "Doors", value: sum((r) => r.doorCount), unit: "count" },
      { id: "windows", label: "Windows", value: sum((r) => r.windowCount), unit: "count" },
    ],
    measurements: [
      {
        id: "floorArea",
        label: "Floor area",
        value: sum((r) => r.floorAreaSqm),
        unit: "sqm",
        meaning: MEASURE_DEFINITIONS.floorArea,
      },
      {
        id: "wallAreaGross",
        label: "Wall area (gross)",
        value: sum((r) => r.wallAreaGrossSqm),
        unit: "sqm",
        meaning: MEASURE_DEFINITIONS.wallAreaGross,
      },
      {
        id: "wallAreaNet",
        label: "Wall area (net)",
        value: sum((r) => r.wallAreaNetSqm),
        unit: "sqm",
        meaning: MEASURE_DEFINITIONS.wallAreaNet,
      },
      {
        id: "perimeter",
        label: "Perimeter",
        value: sum((r) => r.perimeterM),
        unit: "m",
        meaning: MEASURE_DEFINITIONS.perimeter,
      },
      {
        id: "volume",
        label: "Volume",
        // Per room, never total floor area × some single height: rooms have
        // different ceilings, and the whole point of the figure is the air in
        // each of them.
        value: sum((r) => r.floorAreaSqm * r.ceilingHeightM),
        unit: "cbm",
        meaning: MEASURE_DEFINITIONS.volume,
      },
    ],
  };
}

const CUBIC_METRES_TO_CUBIC_FEET = 35.3146667;

/** Cubic metres to cubic feet — the unit dehumidifier ratings are published in. */
export function cubicMetersToCubicFeet(cbm: number): number {
  return cbm * CUBIC_METRES_TO_CUBIC_FEET;
}

File: ios/App/App/Native/ProjectStatistics.swift
import SwiftUI

/// The whole property in numbers.
///
/// The Swift twin of `src/lib/crm/projectStatistics.ts` — same rows, same
/// arithmetic, same two departures from the reference:
///
/// **Volume is promoted.** The reference lists it last, as trivia. Drying is
/// sized from cubic footage under IICRC S500, so here it sits with the
/// measurements an estimate is built from, and it is stated in cubic feet
/// because that is the unit a dehumidifier's coverage is published in.
///
/// **The three ground-surface variants are absent.** The reference reports
/// ground surface "with all walls", "with interior walls" and "without walls".
/// Those need the thickness of every wall and which side belongs to which
/// room. A scan gives wall faces, not wall assemblies. Computing them would
/// mean inventing a thickness and printing the result as a measurement — one
/// honest floor area, defined, beats three fabricated ones.
enum ProjectStats {
    struct Row: Identifiable {
        let id: String
        let label: String
        let value: String
        /// Counts have no definition — "how many doors" needs no defending.
        let meaning: MeasureDefinition?
    }

    static func cubicFeet(_ cubicMetres: Double) -> Double { cubicMetres * 35.314_666_7 }

    /// Guards every figure. A scan held offline, or a room mid-edit, can
    /// arrive with a hole in it, and one bad room must not turn the whole
    /// sheet into "nan sq ft".
    private static func sum(_ rooms: [RoomScan], _ pick: (RoomScan) -> Double) -> Double {
        rooms.reduce(0) { total, room in
            let value = pick(room)
            return total + (value.isFinite ? value : 0)
        }
    }

    /// Gross wall area: perimeter × ceiling height, the same definition the
    /// room sheet and the report already use.
    static func grossWallSqm(_ rooms: [RoomScan]) -> Double {
        sum(rooms) { $0.wallLengthM * $0.ceilingHeightM }
    }

    static func floorAreaSqm(_ rooms: [RoomScan]) -> Double { sum(rooms) { $0.floorAreaSqm } }
    static func perimeterM(_ rooms: [RoomScan]) -> Double { sum(rooms) { $0.wallLengthM } }

    /// Summed PER ROOM. A basement and a stairwell have different ceilings,
    /// and multiplying one blended height by the total floor area describes
    /// neither of them.
    static func volumeCbm(_ rooms: [RoomScan]) -> Double {
        sum(rooms) { $0.floorAreaSqm * $0.ceilingHeightM }
    }

    static func volumeLabel(_ rooms: [RoomScan]) -> String {
        "\(Int(cubicFeet(volumeCbm(rooms)).rounded()))"
    }

    static func summary(_ rooms: [RoomScan]) -> [Row] {
        // Distinct storeys, not rooms-with-a-level: three rooms on the ground
        // floor are one floor.
        let floors = Set(rooms.map(\.level).filter { !$0.isEmpty }).count
        return [
            Row(id: "floors", label: "Floors", value: "\(floors)", meaning: nil),
            Row(id: "rooms", label: "Rooms", value: "\(rooms.count)", meaning: nil),
            Row(id: "doors", label: "Doors",
                value: "\(rooms.reduce(0) { $0 + $1.doorCount })", meaning: nil),
            Row(id: "windows", label: "Windows",
                value: "\(rooms.reduce(0) { $0 + $1.windowCount })", meaning: nil),
            Row(id: "stairs", label: "Stairs",
                value: "\(rooms.reduce(0) { $0 + $1.stairCount })", meaning: nil),
        ]
    }

    static func measurements(_ rooms: [RoomScan]) -> [Row] {
        [
            Row(id: "floorArea", label: "Floor area",
                value: Measure.sqftLabel(floorAreaSqm(rooms)), meaning: .floorArea),
            Row(id: "wallArea", label: "Wall area (gross)",
                value: Measure.sqftLabel(grossWallSqm(rooms)), meaning: .wallArea),
            Row(id: "perimeter", label: "Perimeter",
                value: Measure.ftLabel(perimeterM(rooms)), meaning: .perimeter),
            Row(id: "volume", label: "Volume",
                value: "\(volumeLabel(rooms)) cu ft", meaning: .volume),
        ]
    }
}

/// Three figures and a room count across the top of a project — how big is
/// this job, answered before you have scrolled into any one room.
///
/// Built from `DefinedFigure`, so every number carries its ⓘ and reads the
/// same as the ones on the room sheet.
struct ProjectStatisticsBand: View {
    let rooms: [RoomScan]
    var onSeeAll: () -> Void

    var body: some View {
        VStack(spacing: Brand.Space.hair) {
            Button(action: onSeeAll) {
                SectionHeading(title: "STATISTICS", trailing: "See all")
            }
            .buttonStyle(.plain)

            Card(padding: Brand.Space.small) {
                // Floor area, wall area and volume are what an estimate starts
                // from. Perimeter and the counts live behind `See all` —
                // beyond three figures plus scale, none of them stay readable
                // at this width.
                HStack(alignment: .top, spacing: Brand.Space.small) {
                    DefinedFigure(
                        value: Measure.sqftLabel(ProjectStats.floorAreaSqm(rooms)),
                        unit: nil, meaning: .floorArea)
                    DefinedFigure(
                        value: Measure.sqftLabel(ProjectStats.grossWallSqm(rooms)),
                        unit: "gross", meaning: .wallArea)
                    DefinedFigure(
                        value: ProjectStats.volumeLabel(rooms),
                        unit: "cu ft", meaning: .volume)
                }
            }
        }
    }
}

/// Every figure, with the definition behind it — the reference's own idea,
/// and a good one: an ⓘ on each row means no number is taken on faith.
struct ProjectStatisticsSheet: View {
    let rooms: [RoomScan]
    @Environment(\.dismiss) private var dismiss

    var body: some View {
        NavigationStack {
            List {
                Section("Summary") {
                    ForEach(ProjectStats.summary(rooms)) { StatisticRowView(row: $0) }
                }
                Section("Measurements") {
                    ForEach(ProjectStats.measurements(rooms)) { StatisticRowView(row: $0) }
                }
                Section {
                    Text("""
                        Every measurement here is taken to the wall faces the scan detected. \
                        Where an outline was corrected by hand, the corrected outline is what \
                        was measured.
                        """)
                        .font(.footnote)
                        .foregroundStyle(Brand.inkFaint)
                }
            }
            .navigationTitle("Statistics")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .confirmationAction) {
                    Button("Done") { dismiss() }
                }
            }
        }
    }
}

/// One figure and its ⓘ. Shared with the room-level sheet — the definition
/// popover must read identically wherever a number is shown, which is the
/// whole point of having definitions.
struct StatisticRowView: View {
    let row: ProjectStats.Row
    @State private var showing = false

    var body: some View {
        HStack(spacing: Brand.Space.hair) {
            Text(row.label).foregroundStyle(Brand.ink)
            if row.meaning != nil {
                Image(systemName: "info.circle")
                    .font(.system(size: 12))
                    .foregroundStyle(Brand.blue)
            }
            Spacer()
            Text(row.value)
                .monospacedDigit()
                .foregroundStyle(Brand.inkSoft)
        }
        .contentShape(Rectangle())
        .onTapGesture { if row.meaning != nil { showing = true } }
        .popover(isPresented: $showing) {
            if let meaning = row.meaning {
                VStack(alignment: .leading, spacing: Brand.Space.small) {
                    Text(meaning.title)
                        .font(.system(size: 16, weight: .bold))
                        .foregroundStyle(Brand.ink)
                    Text(meaning.definition)
                        .font(.system(size: 14))
                        .foregroundStyle(Brand.inkSoft)
                        .fixedSize(horizontal: false, vertical: true)
                }
                .padding(Brand.Space.base)
                .frame(width: 300)
                .presentationCompactAdaptation(.popover)
            }
        }
    }
}
